File: engine_data/data/scripts/scene/components/CharacterControllerComponent.cs
using System;
using System.Numerics;
using System.Runtime.CompilerServices;

namespace Unravel.Core
{
    /// <summary>
    /// Provides character controller physics functionality for an entity.
    /// Uses a capsule shape internally with sweep-based movement.
    /// </summary>
    public class CharacterControllerComponent : Component
    {
        /// <summary>
        /// Whether the character is currently touching the ground.
        /// </summary>
        public bool isGrounded
        {
            get { return internal_m2n_cc_get_is_grounded(owner); }
        }

        /// <summary>
        /// Whether the character can currently jump (equivalent to isGrounded).
        /// </summary>
        public bool canJump
        {
            get { return internal_m2n_cc_get_can_jump(owner); }
        }

        /// <summary>
        /// The current velocity of the character controller (read-only, synced from physics).
        /// </summary>
        public Vector3 velocity
        {
            get { return internal_m2n_cc_get_velocity(owner); }
        }

        /// <summary>
        /// The linear velocity of the character controller. Setting this directly overrides the physics velocity.
        /// </summary>
        public Vector3 linearVelocity
        {
            get { return internal_m2n_cc_get_linear_velocity(owner); }
            set { internal_m2n_cc_set_linear_velocity(owner, value); }
        }

        /// <summary>
        /// The radius of the character capsule.
        /// </summary>
        public float radius
        {
            get { return internal_m2n_cc_get_radius(owner); }
            set { internal_m2n_cc_set_radius(owner, value); }
        }

        /// <summary>
        /// The total height of the character capsule.
        /// </summary>
        public float height
        {
            get { return internal_m2n_cc_get_height(owner); }
            set { internal_m2n_cc_set_height(owner, value); }
        }

        /// <summary>
        /// The center offset of the capsule relative to the entity transform.
        /// </summary>
        public Vector3 center
        {
            get { return internal_m2n_cc_get_center(owner); }
            set { internal_m2n_cc_set_center(owner, value); }
        }

        /// <summary>
        /// Maximum height of obstacles the character can step over.
        /// </summary>
        public float stepHeight
        {
            get { return internal_m2n_cc_get_step_height(owner); }
            set { internal_m2n_cc_set_step_height(owner, value); }
        }

        /// <summary>
        /// Maximum slope angle in degrees the character can walk up.
        /// </summary>
        public float slopeLimit
        {
            get { return internal_m2n_cc_get_slope_limit(owner); }
            set { internal_m2n_cc_set_slope_limit(owner, value); }
        }

        /// <summary>
        /// Collision skin width around the character capsule.
        /// </summary>
        public float skinWidth
        {
            get { return internal_m2n_cc_get_skin_width(owner); }
            set { internal_m2n_cc_set_skin_width(owner, value); }
        }

        /// <summary>
        /// Multiplier for world gravity applied to this controller.
        /// </summary>
        public float gravityScale
        {
            get { return internal_m2n_cc_get_gravity_scale(owner); }
            set { internal_m2n_cc_set_gravity_scale(owner, value); }
        }

        /// <summary>
        /// Maximum downward fall speed in m/s. Downward vertical velocity is clamped
        /// to this magnitude each simulation step. Default is 55 (skydiver terminal velocity).
        /// </summary>
        public float terminalVelocity
        {
            get { return internal_m2n_cc_get_terminal_velocity(owner); }
            set { internal_m2n_cc_set_terminal_velocity(owner, value); }
        }

        /// <summary>
        /// Damping applied to linear velocity each step. 0 = no damping, 1 = full damping.
        /// </summary>
        public float linearDamping
        {
            get { return internal_m2n_cc_get_linear_damping(owner); }
            set { internal_m2n_cc_set_linear_damping(owner, value); }
        }

        /// <summary>
        /// Layers that this character should collide with (include filter).
        /// </summary>
        public LayerMask includeLayers
        {
            get { return internal_m2n_cc_get_include_layers(owner); }
            set { internal_m2n_cc_set_include_layers(owner, value); }
        }

        /// <summary>
        /// Layers that this character should ignore (exclude filter).
        /// </summary>
        public LayerMask excludeLayers
        {
            get { return internal_m2n_cc_get_exclude_layers(owner); }
            set { internal_m2n_cc_set_exclude_layers(owner, value); }
        }

        /// <summary>
        /// Effective collision layer mask after applying include and exclude filters.
        /// </summary>
        public LayerMask collisionLayers
        {
            get { return internal_m2n_cc_get_collision_layers(owner); }
        }

        /// <summary>
        /// Moves the character by a displacement vector. The movement is constrained by collisions.
        /// </summary>
        /// <param name="displacement">The world-space displacement to apply.</param>
        public void Move(Vector3 displacement)
        {
            internal_m2n_cc_move(owner, displacement);
        }

        /// <summary>
        /// Makes the character jump straight up with the given initial speed (m/s).
        /// </summary>
        /// <param name="speed">Initial upward speed in m/s. Must be greater than 0.</param>
        public void Jump(float speed)
        {
            internal_m2n_cc_jump(owner, new Vector3(0.0f, speed, 0.0f));
        }

        /// <summary>
        /// Makes the character jump with an arbitrary world-space velocity. The vector's
        /// magnitude is the jump speed and its direction is the jump axis.
        /// </summary>
        /// <param name="velocity">World-space jump velocity. Must be non-zero.</param>
        public void Jump(Vector3 velocity)
        {
            internal_m2n_cc_jump(owner, velocity);
        }

        /// <summary>
        /// Applies an impulse to the character controller.
        /// </summary>
        /// <param name="impulse">The impulse vector to apply.</param>
        public void ApplyImpulse(Vector3 impulse)
        {
            internal_m2n_cc_apply_impulse(owner, impulse);
        }

        /// <summary>
        /// Teleports the character to a new position instantly, bypassing collision detection.
        /// </summary>
        /// <param name="position">The target world position.</param>
        public void Warp(Vector3 position)
        {
            internal_m2n_cc_warp(owner, position);
        }

        // Internal calls
        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_move(Entity eid, Vector3 displacement);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_jump(Entity eid, Vector3 velocity);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_apply_impulse(Entity eid, Vector3 impulse);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_warp(Entity eid, Vector3 position);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern bool internal_m2n_cc_get_is_grounded(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern bool internal_m2n_cc_get_can_jump(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern Vector3 internal_m2n_cc_get_velocity(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern Vector3 internal_m2n_cc_get_linear_velocity(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_linear_velocity(Entity eid, Vector3 velocity);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_radius(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_radius(Entity eid, float radius);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_height(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_height(Entity eid, float height);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern Vector3 internal_m2n_cc_get_center(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_center(Entity eid, Vector3 center);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_step_height(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_step_height(Entity eid, float stepHeight);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_slope_limit(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_slope_limit(Entity eid, float slopeLimit);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_skin_width(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_skin_width(Entity eid, float skinWidth);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_gravity_scale(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_gravity_scale(Entity eid, float scale);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_terminal_velocity(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_terminal_velocity(Entity eid, float speed);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern float internal_m2n_cc_get_linear_damping(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_linear_damping(Entity eid, float damping);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern LayerMask internal_m2n_cc_get_include_layers(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_include_layers(Entity eid, LayerMask mask);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern LayerMask internal_m2n_cc_get_exclude_layers(Entity eid);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern void internal_m2n_cc_set_exclude_layers(Entity eid, LayerMask mask);

        [MethodImpl(MethodImplOptions.InternalCall)]
        private static extern LayerMask internal_m2n_cc_get_collision_layers(Entity eid);
    }
}

File: engine/engine/meta/ecs/components/character_controller_component.cpp
#include "character_controller_component.hpp"
#include <engine/meta/core/math/vector.hpp>
#include <engine/meta/layers/layer_mask.hpp>

#include <serialization/associative_archive.h>
#include <serialization/binary_archive.h>

namespace unravel
{

REFLECT(character_controller_component)
{
    entt::meta_factory<character_controller_component>{}
        .type("character_controller_component"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "character_controller_component"},
            entt::attribute{"category", "PHYSICS"},
            entt::attribute{"pretty_name", "Character Controller"},
        })
        .func<&component_meta<character_controller_component>::exists>("component_exists"_hs)
        .func<&component_meta<character_controller_component>::add>("component_add"_hs)
        .func<&component_meta<character_controller_component>::remove>("component_remove"_hs)
        .func<&component_meta<character_controller_component>::save>("component_save"_hs)
        .func<&component_meta<character_controller_component>::load>("component_load"_hs)
        .data<&character_controller_component::set_radius, &character_controller_component::get_radius>("radius"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "radius"},
            entt::attribute{"pretty_name", "Radius"},
            entt::attribute{"tooltip", "The radius of the character capsule."},
            entt::attribute{"min", 0.01f},
            entt::attribute{"step", 0.1f},
        })
        .data<&character_controller_component::set_height, &character_controller_component::get_height>("height"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "height"},
            entt::attribute{"pretty_name", "Height"},
            entt::attribute{"tooltip", "The total height of the character capsule."},
            entt::attribute{"min", 0.01f},
            entt::attribute{"step", 0.1f},
        })
        .data<&character_controller_component::set_center, &character_controller_component::get_center>("center"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "center"},
            entt::attribute{"pretty_name", "Center"},
            entt::attribute{"tooltip", "The center offset of the capsule relative to the entity transform."},
        })
        .data<&character_controller_component::set_step_height, &character_controller_component::get_step_height>("step_height"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "step_height"},
            entt::attribute{"pretty_name", "Step Offset"},
            entt::attribute{"tooltip", "Maximum height of obstacles the character can step over."},
            entt::attribute{"min", 0.0f},
            entt::attribute{"step", 0.05f},
        })
        .data<&character_controller_component::set_slope_limit, &character_controller_component::get_slope_limit>("slope_limit"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "slope_limit"},
            entt::attribute{"pretty_name", "Slope Limit"},
            entt::attribute{"tooltip", "Maximum slope angle in degrees the character can walk up."},
            entt::attribute{"min", 0.0f},
            entt::attribute{"max", 90.0f},
            entt::attribute{"step", 1.0f},
        })
        .data<&character_controller_component::set_skin_width, &character_controller_component::get_skin_width>("skin_width"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "skin_width"},
            entt::attribute{"pretty_name", "Skin Width"},
            entt::attribute{"tooltip", "Collision skin width around the character capsule."},
            entt::attribute{"min", 0.001f},
            entt::attribute{"step", 0.01f},
        })
        .data<&character_controller_component::set_gravity_scale, &character_controller_component::get_gravity_scale>("gravity_scale"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "gravity_scale"},
            entt::attribute{"pretty_name", "Gravity Scale"},
            entt::attribute{"tooltip", "Multiplier for world gravity applied to this controller."},
            entt::attribute{"step", 0.1f},
        })
        .data<&character_controller_component::set_collision_include_mask, &character_controller_component::get_collision_include_mask>("include_layers"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "include_layers"},
            entt::attribute{"pretty_name", "Include Layers"},
            entt::attribute{"tooltip", "Layers to include when producing collisions."},
        })
        .data<&character_controller_component::set_collision_exclude_mask, &character_controller_component::get_collision_exclude_mask>("exclude_layers"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "exclude_layers"},
            entt::attribute{"pretty_name", "Exclude Layers"},
            entt::attribute{"tooltip", "Layers to exclude when producing collisions."},
        })
        .data<nullptr, &character_controller_component::get_collision_mask>("collision_layers"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "collision_layers"},
            entt::attribute{"pretty_name", "Collision Layers"},
            entt::attribute{"tooltip", "Layers (Include - Exclude) used when producing collisions."},
        })
        .data<&character_controller_component::set_terminal_velocity, &character_controller_component::get_terminal_velocity>("terminal_velocity"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "terminal_velocity"},
            entt::attribute{"pretty_name", "Terminal Velocity"},
            entt::attribute{"tooltip", "Maximum downward fall speed in m/s. Default 55 (skydiver terminal velocity)."},
            entt::attribute{"min", 0.0f},
            entt::attribute{"step", 1.0f},
        })
        .data<&character_controller_component::set_linear_damping, &character_controller_component::get_linear_damping>("linear_damping"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "linear_damping"},
            entt::attribute{"pretty_name", "Linear Damping"},
            entt::attribute{"tooltip", "Damping applied to linear velocity each step. 0 = no damping, 1 = full damping."},
            entt::attribute{"min", 0.0f},
            entt::attribute{"max", 1.0f},
            entt::attribute{"step", 0.05f},
        })
        .data<nullptr, &character_controller_component::is_grounded>("is_grounded"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "is_grounded"},
            entt::attribute{"pretty_name", "Is Grounded"},
        })
        .data<nullptr, &character_controller_component::can_jump>("can_jump"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "can_jump"},
            entt::attribute{"pretty_name", "Can Jump"},
        })
        .data<nullptr, &character_controller_component::get_velocity>("velocity"_hs)
        .custom<entt::attributes>(entt::attributes{
            entt::attribute{"name", "velocity"},
            entt::attribute{"pretty_name", "Velocity"},
        });
}

SAVE(character_controller_component)
{
    try_save(ar, ser20::make_nvp("radius", obj.get_radius()));
    try_save(ar, ser20::make_nvp("height", obj.get_height()));
    try_save(ar, ser20::make_nvp("center", obj.get_center()));
    try_save(ar, ser20::make_nvp("step_height", obj.get_step_height()));
    try_save(ar, ser20::make_nvp("slope_limit", obj.get_slope_limit()));
    try_save(ar, ser20::make_nvp("skin_width", obj.get_skin_width()));
    try_save(ar, ser20::make_nvp("gravity_scale", obj.get_gravity_scale()));
    try_save(ar, ser20::make_nvp("include_layers", obj.get_collision_include_mask()));
    try_save(ar, ser20::make_nvp("exclude_layers", obj.get_collision_exclude_mask()));
    try_save(ar, ser20::make_nvp("terminal_velocity", obj.get_terminal_velocity()));
    try_save(ar, ser20::make_nvp("linear_damping", obj.get_linear_damping()));
}
SAVE_INSTANTIATE(character_controller_component, ser20::oarchive_associative_t);
SAVE_INSTANTIATE(character_controller_component, ser20::oarchive_binary_t);

LOAD(character_controller_component)
{
    float radius{0.5f};
    if(try_load(ar, ser20::make_nvp("radius", radius)))
    {
        obj.set_radius(radius);
    }
    float height{2.0f};
    if(try_load(ar, ser20::make_nvp("height", height)))
    {
        obj.set_height(height);
    }
    math::vec3 center{};
    if(try_load(ar, ser20::make_nvp("center", center)))
    {
        obj.set_center(center);
    }
    float step_height{0.3f};
    if(try_load(ar, ser20::make_nvp("step_height", step_height)))
    {
        obj.set_step_height(step_height);
    }
    float slope_limit{45.0f};
    if(try_load(ar, ser20::make_nvp("slope_limit", slope_limit)))
    {
        obj.set_slope_limit(slope_limit);
    }
    float skin_width{0.08f};
    if(try_load(ar, ser20::make_nvp("skin_width", skin_width)))
    {
        obj.set_skin_width(skin_width);
    }
    float gravity_scale{1.0f};
    if(try_load(ar, ser20::make_nvp("gravity_scale", gravity_scale)))
    {
        obj.set_gravity_scale(gravity_scale);
    }
    layer_mask include_layers;
    if(try_load(ar, ser20::make_nvp("include_layers", include_layers)))
    {
        obj.set_collision_include_mask(include_layers);
    }
    layer_mask exclude_layers;
    if(try_load(ar, ser20::make_nvp("exclude_layers", exclude_layers)))
    {
        obj.set_collision_exclude_mask(exclude_layers);
    }
    float terminal_velocity{55.0f};
    if(try_load(ar, ser20::make_nvp("terminal_velocity", terminal_velocity)) ||
       try_load(ar, ser20::make_nvp("fall_speed", terminal_velocity)))
    {
        obj.set_terminal_velocity(terminal_velocity);
    }
    float linear_damping{0.0f};
    if(try_load(ar, ser20::make_nvp("linear_damping", linear_damping)))
    {
        obj.set_linear_damping(linear_damping);
    }
}

LOAD_INSTANTIATE(character_controller_component, ser20::iarchive_associative_t);
LOAD_INSTANTIATE(character_controller_component, ser20::iarchive_binary_t);

} // namespace unravel
